src/codeu/chat/ServerMain.java

```diff
@@ -31,87 +28,95 @@
-      LOG.error("%s does not exist", persistentPath);
-      System.exit(1);
-    }
+        Uuid id = null;
+        Secret secret = null;
+        int port = -1;
+        // This is the directory where it is safe to store data across runs
+        // of the server.
+        File persistentPath = null;
+        RemoteAddress relayAddress = null;
+
+        try {
+            id = Uuid.parse(args[0]);
+            secret = Secret.parse(args[1]);
+            port = Integer.parseInt(args[2]);
+            persistentPath = new File(args[3]);
+            relayAddress = args.length > 4 ? RemoteAddress.parse(args[4]) : null;
+        } catch (Exception ex) {
+            LOG.error(ex, "Failed to read command arguments");
+            System.exit(1);
+        }
 
-    try (
-        final ConnectionSource serverSource = ServerConnectionSource.forPort(port);
-        final ConnectionSource relaySource = relayAddress == null ? null : new ClientConnectionSource(relayAddress.host, relayAddress.port)
-    ) {
+        if (!persistentPath.isDirectory()) {
+            LOG.error("%s does not exist", persistentPath);
+            System.exit(1);
+        }
 
-      LOG.info("Starting server...");
-      runServer(id, secret, serverSource, relaySource);
+        try (
+                final ConnectionSource serverSource = ServerConnectionSource.forPort(port);
+                final ConnectionSource relaySource = relayAddress == null ? null : new ClientConnectionSource(relayAddress.host, relayAddress.port)
+        ) {
 
-    } catch (IOException ex) {
+            LOG.info("Starting server...");
+            runServer(id, secret, serverSource, relaySource, persistentPath);
 
-      LOG.error(ex, "Failed to establish connections");
+        } catch (IOException ex) {
 
+            LOG.error(ex, "Failed to establish connections");
+
+        }
     }
-  }
 
-  private static void runServer(Uuid id,
-                                Secret secret,
-                                ConnectionSource serverSource,
-                                ConnectionSource relaySource) {
+    private static void runServer(Uuid id,
+                                  Secret secret,
+                                  ConnectionSource serverSource,
+                                  ConnectionSource relaySource,
+                                  File persistentPath) {
 
-    final Relay relay = relaySource == null ?
-                        new NoOpRelay() :
-                        new RemoteRelay(relaySource);
+        final Relay relay = relaySource == null ?
+                new NoOpRelay() :
+                new RemoteRelay(relaySource);
 
-    final Server server = new Server(id, secret, relay);
+        File log = new File(persistentPath, "log.txt");
+        final Server server = new Server(id, secret, relay, log);
 
-    LOG.info("Created server.");
+        LOG.info("Created server.");
 
-    while (true) {
+        server.restore(log);
```

Comment:
do we want to call this here? Imagine we migrated to a database and want to remove this storage solution. We would have to come and edit this file too, to remove this line. Now if this was included in the server constructor, you wouldn't need to worry about this file. the lesser files you put your logic into, the more maintainable it is :)